### PR TITLE
Tlportal 456 - get assignments from Canvas course by course

### PR DESCRIPTION
**Pull request by dlhaines**
Get canvas course assignments from each listed course individually to avoid losing all data due to issues with particular courses.

### CODE REVIEW

File: server/courselist.rb [hunk 1]
```diff
@@ -627,6 +627,22 @@ def run_url_parse_json(new_url)
       logger.debug "#{__method__}: #{__LINE__}: #{new_url}: request_body_ruby: +++#{request_body_ruby.inspect}+++"
       request_body_ruby
     end
+
+    # Get the canvas courses from the list provided by mpathways.
+
+    def getCanvasCourseList(course_data, userid)
+      # could the regexp be a constant?
```
Comment (lsloan) [inline server/courselist.rb]: Could it?  Where would it be defined, in this method or elsewhere?  Would the constant hold the Regexp object?  (Is that even possible?)  Or would the constant hold the regex string?  If you're worried about the time it takes to compile a regex, maybe this instantiation should go somewhere else.  This method could reuse the object instead of recreating it each time the method's invoked.

Comment (lsloan) [inline server/courselist.rb]: Also, I'd give the object a better name than `p`.  I like to use descriptive names, because it will make future maintenance easier.  Plus, we can spare the extra bytes.  :wink:

File: server/courselist.rb [hunk 2]
```diff
@@ -627,6 +627,22 @@ def run_url_parse_json(new_url)
       logger.debug "#{__method__}: #{__LINE__}: #{new_url}: request_body_ruby: +++#{request_body_ruby.inspect}+++"
       request_body_ruby
     end
+
+    # Get the canvas courses from the list provided by mpathways.
```
Comment (lsloan) [inline server/courselist.rb]: Why did you move the code into this method?  It seems to be referenced only in one place, from where it had been extracted.  Are there plans to use this method elsewhere?

File: server/data_provider_canvas_esb.rb [hunk 3]
```diff
@@ -124,38 +124,48 @@ def canvasAPICalendarEventsURL(uniqname, canvas_courses)
...
   end
 
+  # Create a string that expands the array of arguments and concatenates the array prefix to each one
+  def self.urlArrayParameters(array_prefix, parameters)
+    parameters.inject("") { |result, parameter| result << "#{array_prefix}#{parameter}" }
+  end
+
   #  <canvas server>/api/v1/users/self/calendar_events
-  def canvasESBToDoLMS(uniqname, canvas_courses, security_file, esb_application)
+  def canvasESBToDoLMSArray(uniqname, canvas_courses, security_file, esb_application)
```
Comment (lsloan) [inline server/data_provider_canvas_esb.rb]: Does the name of this method mean it returns an array?  It appears to return a `WAPIResultWrapper` object, not an array.

File: server/data_provider_canvas_esb.rb [hunk 4]
```diff
@@ -124,38 +124,48 @@ def canvasAPICalendarEventsURL(uniqname, canvas_courses)
     full_request_url
   end
 
+  # Create a string that expands the array of arguments and concatenates the array prefix to each one
+  def self.urlArrayParameters(array_prefix, parameters)
```
Comment (lsloan) [inline server/data_provider_canvas_esb.rb]: This doesn't seem to be referenced anywhere except for its test, `test_urlArrayParameters`.

File: server/data_provider_canvas_esb.rb [hunk 5]
```diff
@@ -124,38 +124,48 @@ def canvasAPICalendarEventsURL(uniqname, canvas_courses)
...
+    logger.debug "#{self.class.to_s}:#{__method__}:#{__LINE__}: canvas_course >>#{canvas_course}<< r:#{r.inspect}"
     canvas_body = r.result
-    canvas_body_ruby = JSON.parse canvas_body
-
-    return WAPIResultWrapper.new(WAPI::SUCCESS, "got calendar_events from canvas esb", canvas_body_ruby)
-  end
+    logger.debug "#{self.class.to_s}:#{__method__}:#{__LINE__}: canvas_course >>#{canvas_course}<< canvas_body:#{canvas_body.inspect}"
 
+    # if the course access is unauthorized then ignore the course.  This can happen if the student status in the course is inactive.
+    p = Regexp.new(/401 Unauthorized/)
```
Comment (zqian) [inline server/data_provider_canvas_esb.rb]: Did not see this RE p get used anyware.

File: server/spec/test_data_provider_canvas_esb.rb [hunk 6]
```diff
@@ -102,4 +102,15 @@ def test_upcoming_events_studenta
     assert_operator title_0.length, '>', 0, "title has length"
   end
 
+  #### test that the array parameters are expanded correctly.
+  #self.urlArrayParameters('&state[]=',allowed_states)
+  #self.urlArrayParameters
+  #DataProviderCanvasESB
+
+  def test_urlArrayParameters
+    r = DataProviderCanvasESB.urlArrayParameters("&prefix[]=", ["B", "C"])
```
Comment (zqian) [inline server/spec/test_data_provider_canvas_esb.rb]: What is "B" or "C" stand up for? A little documentation would help here.

File: server/spec/test_integration_WAPI_CANVAS.rb [hunk 7]
```diff
@@ -330,7 +275,7 @@ def test_canvas_api_studenta_calendar_events_headers
 
     refute_nil @w
 
-    user='dlhaines'
+    user='XXX'
```
Comment (zqian) [inline server/spec/test_integration_WAPI_CANVAS.rb]: Instead of 'XXX', can we use some meaningful name as a placeholder?

File: server/spec/test_data_provider_canvas_esb.rb (hunk 6 above)
Comment (lsloan) [inline server/spec/test_data_provider_canvas_esb.rb]: Could it be random sample data?

File: server/data_provider_canvas_esb.rb [hunk 8]
```diff
@@ -124,38 +124,48 @@ def canvasAPICalendarEventsURL(uniqname, canvas_courses)
...
-  #   r = @canvasESB_w.get_request "/users/self/upcoming_events?as_user_id=sis_login_id:#{uniqname}"
-  #
-  #   canvas_body = r.result
-  #   canvas_body_ruby = JSON.parse canvas_body
-  #
-  #   return WAPIResultWrapper.new(WAPI::SUCCESS, "got todos from canvas esb", canvas_body_ruby)
-  # end
-
-  # fake list of "canvas_courses"=>["43412", "44525", "44526", "44631", "44630", "44528", "44530"]}
+    canvas_body_ruby = JSON.parse canvas_body
```
Comment (lsloan) [inline server/data_provider_canvas_esb.rb]: I was confused by the name `canvas_body_ruby`.  After I've verified what I thought `JSON.parse` is doing, I think a better name would be `canvas_body_objects`.  Likewise, maybe `canvas_body` should be `canvas_body_json`.

File: server/spec/test_integration_WAPI_CANVAS.rb (hunk 7 above)
Comment (lsloan) [inline server/spec/test_integration_WAPI_CANVAS.rb]: :+1: Try a retired username, like the venerable `bjensen`, or an invalid one, like  `non_existent_user`.

File: server/courselist.rb (hunk 1 above)
Comment (dlhaines) [inline server/courselist.rb]: I'll improve the name and, due to deadlines, make a jira to make the regexp a constant in a later release.

File: server/courselist.rb (hunk 2 above)
Comment (dlhaines) [inline server/courselist.rb]: Having this as a separate method is more understandable and testable.  There is a lot going on in that little method.

File: server/data_provider_canvas_esb.rb (hunk 4 above)
Comment (dlhaines) [inline server/data_provider_canvas_esb.rb]: I'll remove it from this version.

File: server/data_provider_canvas_esb.rb (hunk 3 above)
Comment (dlhaines) [inline server/data_provider_canvas_esb.rb]: I'll improve the name.

File: server/data_provider_canvas_esb.rb (hunk 5 above)
Comment (dlhaines) [inline server/data_provider_canvas_esb.rb]: Nope. It will go away.

File: server/data_provider_canvas_esb.rb (hunk 8 above)
Comment (dlhaines) [inline server/data_provider_canvas_esb.rb]: I'll improve the name of cavas_body_ruby.

File: server/spec/test_data_provider_canvas_esb.rb (hunk 6 above)
Comment (dlhaines) [inline server/spec/test_data_provider_canvas_esb.rb]: They don't stand for anything.  They are just strings for testing.  There is no need to be random.  They can just be arbitrary.  This will also go away when I remove the currently unused urlArrayParameters method.

File: server/spec/test_integration_WAPI_CANVAS.rb (hunk 7 above)
Comment (dlhaines) [inline server/spec/test_integration_WAPI_CANVAS.rb]: Good ideas for future tests.
